File: Stack/Java/Parsing A Boolean Expression.java
// Problem link: https://leetcode.com/problems/parsing-a-boolean-expression/description
// Time: O(n)
// Space: O(n)
class Solution {
    public boolean parseBoolExpr(String expression) {
        Stack<Character> stack = new Stack<>();
        Stack<Character> opr = new Stack<>();
        boolean top;

        for(char chr : expression.toCharArray()) {
            if(chr == ')') {
                if(opr.peek() == '!') {
                    top = stack.pop() == 't' ? true : false;
                    opr.pop();
                    stack.pop();
                    stack.push(top ? 'f' : 't');
                }
                else if(opr.peek() == '|') {
                    top = stack.pop() == 't' ? true : false;
                    opr.pop();

                    while(stack.peek() != '(') {
                        top = top || stack.peek() == 't' ? true : false;
                        stack.pop();
                    }

                    stack.pop();
                    stack.push(top ? 't' : 'f');
                }
                else {
                    top = stack.pop() == 't' ? true : false;
                    opr.pop();

                    while(stack.peek() != '(') {
                        top = top && stack.peek() == 't' ? true : false;
                        stack.pop();
                    }

                    stack.pop();
                    stack.push(top ? 't' : 'f');
                }
            }
            else {
                if(chr == '!' || chr == '|' || chr == '&') {
                    opr.push(chr);
                }
                else if(chr != ',') {
                    stack.push(chr);
                }
            }
        }

        return stack.pop() == 't' ? true : false;
    }
}

File: Stack/Python/Parsing A Boolean Expression.py
# Problem link: https://leetcode.com/problems/parsing-a-boolean-expression/description
# Time: O(n)
# Space: O(n)
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        store = {'f' : False, 't' : True, False: 'f', True : 't'}
        stack, opr = [], []

        for char in expression:
            if char == ')':
                if opr[-1] == '!':
                    top = store[stack.pop()]
                    stack.pop()
                    opr.pop()
                    stack.append(store[not top])
                elif opr[-1] == '|':
                    top = store[stack.pop()]
                    opr.pop()

                    while stack and stack[-1] != '(':
                        top = top or (store[stack[-1]])
                        stack.pop()
                    
                    stack.pop()
                    stack.append(store[top])
                else:
                    top = store[stack.pop()]
                    opr.pop()

                    while stack[-1] != '(':
                        top = top and (store[stack[-1]])
                        stack.pop()
                    
                    stack.pop()
                    stack.append(store[top])
            else:
                if char == '&' or char == '|' or char == '!':
                    opr.append(char)
                elif char != ',':
                    stack.append(char)
                    
        return store[stack.pop()]
